File: Zeze/Raft/RocksRaft/CollSet1.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Zeze.Serialize;

namespace Zeze.Raft.RocksRaft
{
	public class CollSet1<V> : CollSet<V>
	{
		public override bool Add(V item)
        {
			if (IsManaged)
			{
				var setlog = (LogSet1<V>)Transaction.Current.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				return setlog.Add(item);
			}
			else
			{
				var newset = _set.Add(item);
				if (newset == _set)
					return false;
				_set = newset;
				return true;
			}
		}

		public override bool Remove(V item)
		{
			if (IsManaged)
			{
				var setlog = (LogSet1<V>)Transaction.Current.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				return setlog.Remove(item);
			}
			else
			{
				var newset = _set.Remove(item);
				if (newset == _set)
					return false;
				_set = newset;
				return true;
			}
		}

		public override void Clear()
		{
			if (IsManaged)
			{
				var setlog = (LogSet1<V>)Transaction.Current.LogGetOrAdd(Parent.ObjectId + VariableId, CreateLogBean);
				setlog.Clear();
			}
			else
			{
				_set = _set.Clear();
			}
		}

		public override LogBean CreateLogBean()
		{
			var log = new LogSet1<V>();
			log.Belong = Parent;
			log.This = this;
			log.VariableId = VariableId;
			log.Value = _set;
			return log;
		}

		public override void FollowerApply(Log _log)
		{
			var log = (LogSet1<V>)_log;
			var tmp = _set;
			tmp = tmp.Union(log.Added);
			tmp = tmp.Except(log.Removed);
			_set = tmp;
		}

		public override void LeaderApplyNoRecursive(Log _log)
		{
			var log = (LogSet1<V>)_log;
			_set = log.Value;
		}

        protected override void InitChildrenRootInfo(Record.RootInfo root)
        {
        }

		public override Bean CopyBean()
		{
			var copy = new CollSet1<V>();
			copy._set = _set;
			return copy;
		}
	}
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Raft/RocksRaft/CollSet1.java
package Zeze.Raft.RocksRaft;

import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.IByteBuffer;
import Zeze.Serialize.SerializeHelper;
import Zeze.Util.Reflect;

public class CollSet1<V> extends CollSet<V> {
	protected final SerializeHelper.CodecFuncs<V> valueCodecFuncs;
	private final int logTypeId;

	public CollSet1(Class<V> valueClass) {
		valueCodecFuncs = SerializeHelper.createCodec(valueClass);
		logTypeId = Zeze.Transaction.Bean.hash32("Zeze.Raft.RocksRaft.LogSet1<" + Reflect.getStableName(valueClass) + '>');
	}

	private CollSet1(int logTypeId, SerializeHelper.CodecFuncs<V> valueCodecFuncs) {
		this.valueCodecFuncs = valueCodecFuncs;
		this.logTypeId = logTypeId;
	}

	@Override
	public boolean add(V item) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var setLog = (LogSet1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			return setLog.add(item);
		}
		var newSet = set.plus(item);
		if (newSet == set)
			return false;
		set = newSet;
		return true;
	}

	@Override
	public boolean remove(V item) {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var setLog = (LogSet1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			return setLog.remove(item);
		}
		var newSet = set.minus(item);
		if (newSet == set)
			return false;
		set = newSet;
		return true;
	}

	@Override
	public void clear() {
		if (isManaged()) {
			@SuppressWarnings("unchecked")
			var setLog = (LogSet1<V>)Transaction.getCurrent().logGetOrAdd(
					parent().objectId() + variableId(), this::createLogBean);
			setLog.clear();
		} else
			set = org.pcollections.Empty.set();
	}

	@Override
	public LogBean createLogBean() {
		var log = new LogSet1<>(logTypeId, valueCodecFuncs);
		log.setBelong(parent());
		log.setThis(this);
		log.setVariableId(variableId());
		log.setValue(set);
		return log;
	}

	@Override
	public void followerApply(Log _log) {
		@SuppressWarnings("unchecked")
		var log = (LogSet1<V>)_log;
		set = set.plusAll(log.getAdded()).minusAll(log.getRemoved());
	}

	@SuppressWarnings("unchecked")
	@Override
	public void leaderApplyNoRecursive(Log _log) {
		set = ((LogSet1<V>)_log).getValue();
	}

	@Override
	protected void initChildrenRootInfo(Record.RootInfo root) {
	}

	@Override
	public CollSet1<V> copy() {
		var copy = new CollSet1<>(logTypeId, valueCodecFuncs);
		copy.set = set;
		return copy;
	}

	@Override
	public void encode(ByteBuffer bb) {
		var tmp = getSet();
		bb.WriteUInt(tmp.size());
		var encoder = valueCodecFuncs.encoder;
		for (var e : tmp)
			encoder.accept(bb, e);
	}

	@Override
	public void decode(IByteBuffer bb) {
		clear();
		var decoder = valueCodecFuncs.decoder;
		for (int i = bb.ReadUInt(); i > 0; i--)
			add(decoder.apply(bb));
	}
}
